### DISCUSSION TITLE
libpng not found

### DISCUSSION THREAD

**Discussion by Penguinlay** [Homebrew/discussions#1044] [Everyday usage] [answered]
`brew info libpng`
```
libpng: stable 1.6.37 (bottled), HEAD
Library for manipulating PNG images
http://www.libpng.org/pub/png/libpng.html
/opt/homebrew/Cellar/libpng/1.6.37 (27 files, 1.3MB) *
  Poured from bottle on 2021-03-13 at 19:54:05
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/libpng.rb
License: libpng-2.0
==> Options
--HEAD
	Install HEAD version
==> Analytics
install: 133,331 (30 days), 391,160 (90 days), 1,575,926 (365 days)
install-on-request: 2,774 (30 days), 8,091 (90 days), 35,184 (365 days)
build-error: 0 (30 days)
```

`which libpng`
```
libpng not found
```

It seems that in `/opt/homebrew/bin`:
```
lrwxr-xr-x  1 shein  admin    34 Mar 13 19:14 libpng -> /opt/homebrew/Cellar/libpng/1.6.37
```
where libpng is linked to the directory `1.6.37` instead of an executable.

As a result, during `yarn install` on Gatsby project, I am getting:
```
error /Users/shein/kk/gh/bvr/node_modules/pngquant-bin: Command failed.
Exit code: 1
Command: node lib/install.js
Arguments:
Directory: /Users/shein/kk/gh/bvr/node_modules/pngquant-bin
Output:
⚠ Cannot find module '/Users/shein/kk/gh/bvr/node_modules/@sindresorhus/is/dist/index.js'. Please verify that the package.json has a valid "main" entry
  ⚠ pngquant pre-build test failed
  ℹ compiling from source
  ✖ Error: pngquant failed to build, make sure that libpng is installed
```

My `brew config` is:
```
HOMEBREW_VERSION: 3.0.5
ORIGIN: https://github.com/Homebrew/brew
HEAD: 8b677c0ddd5f96b5521c1b353a9ca07dfe57b80e
Last commit: 5 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 6ed3d5916f86d85930985ff22b659644976a12b9
Core tap last commit: 46 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 12.0 build 1200
Git: 2.24.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.3-arm64
CLT: 12.0.32.29
Xcode: 12.5 => /Applications/Xcode-beta.app/Contents/Developer
Rosetta 2: false
```

`brew list`
```
aom
autoconf
automake
cairo
cfitsio
docbook
docbook-xsl
fftw
fontconfig
freetype
fribidi
gcc
gdbm
gdk-pixbuf
gettext
ghostscript
giflib
glib
gmp
gnu-getopt
gobject-introspection
graphite2
harfbuzz
hdf5
hwloc
icu4c
ilmbase
imagemagick
isl
jpeg
libde265
libevent
libexif
libffi
libgsf
libheif
libidn2
libimagequant
liblqr
libmatio
libmpc
libomp
libpng
libpthread-stubs
librsvg
libspng
libtiff
libtool
libunistring
libx11
libxau
libxcb
libxdmcp
libxext
libxml2
libxrender
little-cms2
lzo
micro
mozjpeg
mpdecimal
mpfr
nspr
nss
open-mpi
openexr
openjpeg
openslide
openssl@1.1
orc
pango
pcre
pixman
pkg-config
pngquant
poppler
python@3.9
qt@5
readline
shared-mime-info
sqlite
szip
tcl-tk
vips
webp
wget
x265
xmlto
xorgproto
xz
```

**Comment by carlocab** [comment]
`libpng` is a library, so you shouldn't expect it to be an executable. It does, however, ship the executables `libpng-config`, `libpng16-config`, `png-fix-itxt`, and `pngfix`. Thus, `which libpng` _should_ return `libpng not found`. (If it doesn't, something's broken.)

>It seems that in `/opt/homebrew/bin`

`libpng` should not be in `/opt/homebrew/bin`. Are you sure you're not looking in `/opt/homebrew/opt`? You should also see something like this (among other entries) in `/opt/homebrew/lib`:
```
Permissions Size User       Date Modified Name
lrwxr-xr-x    10 carlocab   14 Apr  2019  libpng.a -> libpng16.a
lrwxr-xr-x    14 carlocab   14 Apr  2019  libpng.dylib -> libpng16.dylib
.rw-r--r--  205k carlocab   17 Nov  2020  libpng16.16.dylib
.rw-r--r--  251k carlocab   14 Apr  2019  libpng16.a
lrwxr-xr-x    17 carlocab   14 Apr  2019  libpng16.dylib -> libpng16.16.dylib
drwxr-xr-x     - carlocab   17 Nov  2020  pkgconfig
```

**Reply by Penguinlay** [reply]
In that case, most likely the issue is with one of the node packages.

**Reply by Penguinlay** [reply]
One thing I am sure is something that changed in the past couple of days outside node packages on Mac M1 makes thing breaking. I could run the Gatsby starter a few days ago but not anymore with the same above error now. So, I am going through the packages installed through Homebrew at the moment to see which breaks the build.

**Comment by carlocab** [comment, marked as answer]
I don't use Node, so I don't think I can be much more helpful here. However, the following error
```
Cannot find module '/Users/shein/kk/gh/bvr/node_modules/@sindresorhus/is/dist/index.js'. Please verify that the package.json has a valid "main" entry
```
makes me think one of your Node modules or Yarn packages is broken. I suspect that that's what changed, not something that you installed with `brew`.

@derrabus might be able to help you here if they have the time to take a look.

**Reply by Penguinlay** [reply]
I tried with `npm install --force` after deleting `node-modules` and `yarn.lock`. And, it works now on my Mac with M1 chip. By the way, the site is hosted on Netlify which I believe uses linux and yarn works fine there. So, something wrong with one of the dependencies of yarn on mac m1, I think.

**Reply by carlocab** [reply]
Possibly. It could also be because you're running a beta version of macOS or that you have an outdated CLT (or both):
```
macOS: 11.3-arm64
CLT: 12.0.32.29
```

**Reply by derrabus** [reply]
The error message looks like an NPM package needs libpng for compiling a native node addon. I don‘t think I can be much of a help here because I hardly use native packages, sorry.

**Reply by Penguinlay** [reply]
It's alright. Thank you both for replying. For now, I just installed those packages forcefully with npm as I am not currently using any image optimization currently at the site's beta stage.
